lib/Dialect/TensorExt/Transforms/RotateAndReduce.cpp

```diff
@@ -0,0 +1,196 @@
+                  return failure();
+                }
+                LLVM_DEBUG(llvm::dbgs()
+                           << "Adding valid index " << accessIndex << "\n");
+                accessIndices.insert(accessIndex);
+                return success();
+              })
+              .Default([&](Operation *op) { return failure(); });
+
+      if (failed(result)) {
+        return;
+      }
+    }
+
+    // The test for a match is now: does the number of accessed indices exactly
+    // match the size of the tensor? I.e., does each tensor element show up
+    // exactly once in the reduction?
+    auto tensorShape =
+        inputTensors.begin()->getType().cast<RankedTensorType>().getShape();
+    if (tensorShape.size() != 1 || tensorShape[0] != accessIndices.size()) {
+      LLVM_DEBUG(llvm::dbgs()
+                 << "Not replacing op because tensor shape ("
+                 << inputTensors.begin()->getType()
+                 << ") is not fully reduced. Only " << accessIndices.size()
+                 << " of " << tensorShape[0] << " indices were accessed\n");
+      return;
+    }
+
+    // From here we know we will succeed.
+    auto b = ImplicitLocOpBuilder(op->getLoc(), op);
+    Value inputTensor = *inputTensors.begin();
+    Operation *finalOp;
+    for (int64_t shiftSize = tensorShape[0] / 2; shiftSize > 0;
+         shiftSize /= 2) {
+      auto rotatedTensor = b.create<tensor_ext::RotateOp>(
+          inputTensor, b.create<arith::ConstantOp>(b.getIndexAttr(shiftSize)));
+      // FIXME: replace AddIOp with templated op
+      auto addOp = b.create<arith::AddIOp>(inputTensor, rotatedTensor);
+      finalOp = addOp;
+      inputTensor = addOp.getResult();
+    }
+
+    auto *parentOp = op->getParentOp();
+    // We can extract at any index; every index contains the same reduced value.
+    auto extractOp = b.create<tensor::ExtractOp>(
+        finalOp->getResult(0), b.create<arith::ConstantIndexOp>(0).getResult());
+    op->replaceAllUsesWith(extractOp);
+    LLVM_DEBUG(llvm::dbgs() << "Post-replacement: " << *parentOp << "\n");
+
+    // Mark all ops in the reduction as visited so we don't try to replace them
+    // twice.
+    for (Operation *visitedOp : visitedReductionOps) {
+      visited.insert(visitedOp);
+    }
+  }
+
+  void runOnOperation() override {
+    DenseSet<llvm::StringRef> supportedOps = {"arith.addi", "arith.muli"};
+    DenseSet<Operation *> visited;
+    // Traverse the IR in reverse order so that we can eagerly compute backward
+    // slices for each operation.
+    // FIXME: j2kun wonders: is traversing in forward order and computing
+    // backward slices at each step actually bad performance-wise? I need to
+    // test this in isolation. Also wonder: if there's no matched pattern, is
+    // traversing backward and repeatedly computing the same backward slice
+    // going to be too slow? If so, refactor into an analysis pass that caches
+    // the slice at each op.
```

Comment:
Is there even a way around worst-case O(n^2) here with the current constraints? If there was an addition over all values except one from a vector of length n, for example.

Reply:
I think what I'd have to do is define a custom dataflow analysis that associates each op in the IR to the multiset of tensor indices accessed by its backward slice. I.e., each op would come with a set

```
(t_1, [i_{1,1}, i_{1,2}, ..., i_{1, k_1}],
(t_2, [i_{2,1}, i_{2,2}, ..., i_{2, k_2}],
...
(t_n, [i_{n,1}, i_{n,2}, ..., i_{n, k_n}],
```

Where `t_i` is a tensor and the list is the set of indices accessed by ops in the backward slice. Then of course the join operation in the dataflow lattice is a particular sort of set union, and we could impose an artificial cap of `TOO_MANY_TENSORS` once a join causes it to combine two tensors, so that the dataflow size stays small (though without an intentional data structure, it will be O(n^2) space).

Then we could just do a linear pass over the ops until we find index sets that are sufficiently large as to benefit from a rotate-and-reduce, and replace the ops.

Actually... I wish I had thought of this first. This idea is not that hard to implement. I kept thinking of having an analysis pass that cached backward slices, but this is much cleaner. We could even have the analysis templated by the op type, and instantiate one for each of `addi`, `muli`.

Reply:
I will check this in tonight as a reference and do try this tomorrow.